ndk/src/midi.rs

```diff
@@ -0,0 +1,300 @@
+        }
+    }
+
+    /// Gets the number of output (receiving) ports available on the specified MIDI device.
+    pub fn num_output_ports(&self) -> Result<usize> {
+        let num_output_ports = unsafe { ffi::AMidiDevice_getNumOutputPorts(self.ptr.as_ptr()) };
+        if num_output_ports >= 0 {
+            Ok(num_output_ports as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_output_ports as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+
+    /// Gets the MIDI device type.
+    pub fn device_type(&self) -> Result<MidiDeviceType> {
+        let device_type = unsafe { ffi::AMidiDevice_getType(self.ptr.as_ptr()) };
+        if device_type >= 0 {
+            let device_type = MidiDeviceType::try_from(device_type as u32).map_err(|e| {
+                NdkMediaError::UnknownResult(ffi::media_status_t(e.number as c_int))
+            })?;
+            Ok(device_type)
+        } else {
+            Err(NdkMediaError::from_status(ffi::media_status_t(device_type)).unwrap_err())
+        }
+    }
+
+    /// Opens the input port so that the client can send data to it.
+    pub fn open_input_port(&self, port_number: i32) -> Result<MidiInputPort> {
+        unsafe {
+            let input_port =
+                construct(|res| ffi::AMidiInputPort_open(self.ptr.as_ptr(), port_number, res))?;
+            Ok(MidiInputPort::from_ptr(NonNull::new_unchecked(input_port)))
+        }
+    }
+
+    /// Opens the output port so that the client can receive data from it.
+    pub fn open_output_port(&self, port_number: i32) -> Result<MidiOutputPort> {
+        unsafe {
+            let output_port =
+                construct(|res| ffi::AMidiOutputPort_open(self.ptr.as_ptr(), port_number, res))?;
+            Ok(MidiOutputPort::from_ptr(NonNull::new_unchecked(
+                output_port,
+            )))
+        }
+    }
+}
+
+impl Drop for MidiDevice {
+    fn drop(&mut self) {
+        let status = unsafe { ffi::AMidiDevice_release(self.ptr.as_ptr()) };
+        NdkMediaError::from_status(status).unwrap();
+    }
+}
+
+pub struct MidiInputPort<'a> {
+    ptr: NonNull<ffi::AMidiInputPort>,
+    _marker: PhantomData<&'a MidiDevice>,
+}
+
+impl<'a> fmt::Debug for MidiInputPort<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MidiInputPort")
+            .field("inner", &self.ptr)
+            .finish()
+    }
+}
+
+impl<'a> MidiInputPort<'a> {
+    /// Assumes ownership of `ptr`
+    ///
+    /// # Safety
+    /// `ptr` must be a valid pointer to an Android [`ffi::AMidiInputPort`].
+    pub unsafe fn from_ptr(ptr: NonNull<ffi::AMidiInputPort>) -> Self {
+        Self {
+            ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn ptr(&self) -> NonNull<ffi::AMidiInputPort> {
+        self.ptr
+    }
+
+    /// Sends data to the specified input port.
```

Comment:
Nit (here and below): I know this documentation was inherited from Android, but `specified` is a bit vague here since you can only "specify" `self` :)

Reply:
aa26247 
Fixed!